docs/book/common_code/machine_controller.md

```diff
@@ -0,0 +1,265 @@
+}
+```
+{% endmethod %}
+
+{% method %}
+## MachineSpec
+
+The `ProviderSpec` is recommended to be a serialized API object in a format 
+owned by that provider. This will allow the configuration to be strongly typed,
+versioned, and have as much nested depth as appropriate. These provider-specific
+API definitions are meant to live outside of the Machine API, which will allow
+them to evolve independently of it. Attributes like instance type, which 
+network to use, and the OS image all belong in the `ProviderSpec`.
+
+Some providers and tooling depend on an annotation to be set on the `Machine`
+to determine if provisioning has completed. For example, the `clusterctl`
+command does this [here](https://github.com/kubernetes-sigs/cluster-api/blob/a30de81123009a5f91ade870008c1a35f7ce4b35/cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go#L555):
+```go
+		// TODO: update once machine controllers have a way to indicate a machine has been provisoned. https://github.com/kubernetes-sigs/cluster-api/issues/253
+		// Seeing a node cannot be purely relied upon because the provisioned master will not be registering with
+		// the stack that provisions it.
+		ready := m.Status.NodeRef != nil || len(m.Annotations) > 0
+		return ready, nil
+```
+
+{% sample lang="go" %}
+```go
+// MachineSpec defines the desired state of Machine
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The full, authoritative list of taints to apply to the corresponding
+	// Node. This list will overwrite any modifications made to the Node on
+	// an ongoing basis.
+	// +optional
+	Taints []corev1.Taint `json:"taints,omitempty"`
+
+	// Provider-specific configuration to use during node creation.
+	// +optional
+	ProviderSpec ProviderSpec `json:"providerSpec"`
+
+	// Versions of key software to use. This field is optional at cluster
+	// creation time, and omitting the field indicates that the cluster
+	// installation tool should select defaults for the user. These
+	// defaults may differ based on the cluster installer, but the tool
+	// should populate the values it uses when persisting Machine objects.
+	// A Machine spec missing this field at runtime is invalid.
+	// +optional
+	Versions MachineVersionInfo `json:"versions,omitempty"`
+
+	// To populate in the associated Node for dynamic kubelet config. This
+	// field already exists in Node, so any updates to it in the Machine
+	// spec will be automatially copied to the linked NodeRef from the
+	// status. The rest of dynamic kubelet config support should then work
+	// as-is.
+	// +optional
+	ConfigSource *corev1.NodeConfigSource `json:"configSource,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## MachineStatus
+
+Note that the **NodeRef** field may not exist, in particular in the
```

Comment:
I'd recommend against saying 'may not exist' and instead say that 'it may be unused' or 'it may not be set'